experimental/sample_web_static/main.c

```diff
@@ -0,0 +1,187 @@
+  }
+
+  if (iree_status_is_ok(status)) {
+    status = create_bytecode_module(&state->module);
+  }
+  if (iree_status_is_ok(status)) {
+    status = iree_runtime_session_append_module(state->session, state->module);
+  }
+
+  const char kMainFunctionName[] = "module.predict";
+  if (iree_status_is_ok(status)) {
+    status = iree_runtime_call_initialize_by_name(
+        state->session, iree_make_cstring_view(kMainFunctionName),
+        &state->call);
+  }
+
+  if (!iree_status_is_ok(status)) {
+    iree_status_fprint(stderr, status);
+    iree_status_free(status);
+    free(state);
+    return NULL;
+  }
+
+  return state;
+}
+
+void cleanup_sample(iree_sample_state_t* state) {
+  iree_runtime_call_deinitialize(&state->call);
+
+  // Cleanup session and instance.
+  iree_hal_device_release(state->device);
+  iree_runtime_session_release(state->session);
+  iree_runtime_instance_release(state->instance);
+  iree_vm_module_release(state->module);
+
+  free(state);
+}
+
+int run_sample(iree_sample_state_t* state, float* image_data) {
+  iree_status_t status = iree_ok_status();
+
+  iree_runtime_call_reset(&state->call);
+
+  iree_hal_buffer_view_t* arg_buffer_view = NULL;
+  iree_hal_dim_t buffer_shape[] = {1, 28, 28, 1};
+  iree_hal_memory_type_t input_memory_type =
+      IREE_HAL_MEMORY_TYPE_HOST_LOCAL | IREE_HAL_MEMORY_TYPE_DEVICE_VISIBLE;
+  if (iree_status_is_ok(status)) {
+    status = iree_hal_buffer_view_allocate_buffer(
+        iree_hal_device_allocator(state->device), buffer_shape,
+        IREE_ARRAYSIZE(buffer_shape), IREE_HAL_ELEMENT_TYPE_FLOAT_32,
+        IREE_HAL_ENCODING_TYPE_DENSE_ROW_MAJOR, input_memory_type,
+        IREE_HAL_BUFFER_USAGE_DISPATCH | IREE_HAL_BUFFER_USAGE_TRANSFER,
+        iree_make_const_byte_span((void*)image_data, sizeof(float) * 28 * 28),
+        &arg_buffer_view);
+  }
+  if (iree_status_is_ok(status)) {
+    status = iree_runtime_call_inputs_push_back_buffer_view(&state->call,
+                                                            arg_buffer_view);
+  }
+  iree_hal_buffer_view_release(arg_buffer_view);
+
+  if (iree_status_is_ok(status)) {
+    status = iree_runtime_call_invoke(&state->call, /*flags=*/0);
+  }
+
+  // Get the result buffers from the invocation.
+  iree_hal_buffer_view_t* ret_buffer_view = NULL;
+  if (iree_status_is_ok(status)) {
+    status = iree_runtime_call_outputs_pop_front_buffer_view(&state->call,
+                                                             &ret_buffer_view);
+  }
+
+  // Read back the results. The output of the mnist model is a 1x10 prediction
+  // confidence values for each digit in [0, 9].
+  float predictions[1 * 10] = {0.0f};
+  if (iree_status_is_ok(status)) {
+    iree_hal_buffer_read_data(iree_hal_buffer_view_buffer(ret_buffer_view), 0,
+                              predictions, sizeof(predictions));
+  }
+  iree_hal_buffer_view_release(ret_buffer_view);
+
+  if (iree_status_is_ok(status)) {
```

Comment:
could reduce nesting here by checking for failure and returning and then taking the success out to the top level
```suggestion
  if (!iree_status_is_ok(status)) {
     // fail
    return -1;
  }
  // success
  return result_idx;
```